src/features/QuickAdd/QuickAdd.tsx

```diff
@@ -1,19 +1,52 @@
+'use client';
+
 import { Modal } from '@/components/Modal/Modal';
+import { getProduct } from '@/features/QuickAdd/transforms';
+import { QuickAddQueryResponse, QuickAddQueryVariables } from '@/types/takeshape';
+import { QueryReference, useLoadableQuery, useReadQuery } from '@apollo/client';
+import { useAtomValue } from 'jotai';
+import { useResetAtom } from 'jotai/utils';
+import { Suspense, useEffect } from 'react';
 import { QuickAddItem } from './components/QuickAddItem';
 import { QuickAddItemLoading } from './components/QuickAddItemLoading';
-import { QuickAddProduct } from './types';
+import { QuickAddQuery } from './queries';
+import { quickAddAtom } from './store';
 
-export type QuickAddProps = {
-  isOpen: boolean;
+type ReadQuickAddProps = {
   onClose: () => void;
-  product: QuickAddProduct | null;
+  productQueryRef: QueryReference<QuickAddQueryResponse, QuickAddQueryVariables>;
 };
 
-export const QuickAdd = ({ isOpen, onClose, product }: QuickAddProps) => {
+function ReadQuickAdd({ onClose, productQueryRef }: ReadQuickAddProps) {
+  const { data } = useReadQuery(productQueryRef);
+  const product = getProduct(data);
+
+  if (!product) {
+    throw new Error('Error loading QuickAdd item, try again');
+  }
+
+  return <QuickAddItem product={product} onClose={onClose} />;
+}
+
+export const QuickAdd = () => {
+  const quickAdd = useAtomValue(quickAddAtom);
+  const resetQuickAdd = useResetAtom(quickAddAtom);
+
+  const [loadProduct, productQueryRef] = useLoadableQuery<QuickAddQueryResponse, QuickAddQueryVariables>(QuickAddQuery);
+
+  useEffect(() => {
+    if (quickAdd?.productHandle) {
+      void loadProduct({
+        handle: quickAdd.productHandle
+      });
+    }
+  }, [loadProduct, quickAdd]);
+
   return (
-    <Modal isOpen={isOpen} onClose={onClose} showCloseButton={true}>
-      {!product && <QuickAddItemLoading />}
-      {product && <QuickAddItem product={product} onClose={onClose} />}
+    <Modal isOpen={Boolean(quickAdd)} onClose={() => resetQuickAdd()} showCloseButton={true}>
+      <Suspense fallback={<QuickAddItemLoading />}>
```

Comment:
I wanted to take advantage of the new Suspense capabilities somewhere. It offers a cleaner API for handling loading, and Apollo / Next can do some streaming data magic. Works well, but requires a slightly different way of thinking about queries.